File: TNG/source/create_mock.py
"""
Load text data like this

---
# redshift: 0.2 - 12.0
# tan_LoS: 0.06 0.10
# area: 5400.0 arcsec x 5400.0 arcsec
# theta_x[arcsec] theta_y[arcsec] redshift log_lumi_dis vz[km/s] rmass_half log_sfr log_mstar log_Z
5.062337e+03 2.392666e+02 0.1893 27.4628 2.098443e+02 1.197039e+00 0.7396 10.3443 -1.7383
---

"""

cspeed = 2.9979e10  # speed of light in cm/s
Jy = 1.0e-23        # jansky (erg/s/cm2/Hz)
arcsec = 4.848136811094e-6 # [rad] ... arcmin / 60 //


import numpy as np
import os
import argparse

import pandas as pd

import h5py

from line_model import calc_line_luminosity, line_dict

parser = argparse.ArgumentParser(description="Create mock data for SUBLIME")

parser.add_argument("--seed", type=int, default=123, help="random seed")

### I/O parameters
parser.add_argument("--path", type=str, default="./output/lightcone.txt")
parser.add_argument("--output_dir", type=str, default="output")

### area and angular resolution
parser.add_argument("--side_length", type=float, default=100.0, help="side length in arcsec")
parser.add_argument("--angular_resolution", type=float, default=30, help="angular resolution in arcsec")

### spectral range and resolution
parser.add_argument("--fmin", type=float, default=10.0, help="minimum frequency in GHz")
parser.add_argument("--fmax", type=float, default=100.0, help="maximum frequency in GHz")
parser.add_argument("--R", type=float, default=100, help="spectral resolution R")

### line model paramters 
parser.add_argument("--z_dependence_alpha", type=float, default=1.0, help="The SFR at very high redshift is multiplied by this alpha.")
parser.add_argument("--z_middle", type=float, default=8.0, help="If z_dependence_alpha != 1, the SFR deviates from the original value as redshift becomes higher; z_middle is the redshift when the factor is 0.5(alpha + 1)")
parser.add_argument("--sigma", type=float, default=0.2, help="The scatter in log [dex] added to luminosity - SFR relation")

### galaxy catalog parameters
parser.add_argument("--threshold", type=float, default=-1, help="intensity map is generated when threshold < 0. If threshold > 0, galaxy catalogs with flux [erg/s/cm2] > threshold is generated.")

args = parser.parse_args()

np.random.seed(args.seed)   

print("# frequency: {:.4f} - {:.4f} [GHz]".format(args.fmin, args.fmax))
print("# spectral resolution R: {:.4f}".format(args.R))
print("# area : {:.4f} arcsec x {:.4f} arcsec".format(args.side_length, args.side_length))
print("# angular resolution : {:.4f} arcsec".format(args.angular_resolution))

if args.threshold < 0:
    print("# Create intensity map")
else:
    print("# Create galaxy number density map with flux > {:e} erg/s/cm2".format(args.threshold))

### Define the spectral bins
flist = []
dflist = []

fnow = args.fmin * 1e9
while fnow <= args.fmax * 1e9:
    flist.append(fnow)
    dflist.append(fnow / args.R)
    fnow += fnow / args.R
flist = np.array(flist, dtype=np.float32)
dflist = np.array(dflist, dtype=np.float32)
print(f"frequency range is {flist[0]/1e9} - {flist[-1]/1e9} [GHz]")

### Load data
print("# Load data from {}".format(args.path))

df = pd.read_csv(args.path, comment="#", header=None, delim_whitespace=True, dtype=np.float32)
lc_data = df.values
x = lc_data[:, 0]
y = lc_data[:, 1]
z = lc_data[:, 2]
log_lumi_dis = lc_data[:, 3]
vz = lc_data[:, 4] * 1e5 # [cm/s]
rmass_half = lc_data[:, 5]
log_sfr = lc_data[:, 6]
log_mstar = lc_data[:, 7]
log_Z = lc_data[:, 8]

ix = np.floor(x / args.angular_resolution).astype(np.int32)
iy = np.floor(y / args.angular_resolution).astype(np.int32)

print("# Number of galaxies: {}".format(len(x)))

z_obs = z + vz / cspeed * (1 + z)
sfr_factor = (args.z_dependence_alpha - 1) * (0.5 * ( np.tanh( 0.5 * (z - args.z_middle) ) + 1 ) ) + 1
log_sfr += np.log10(sfr_factor)
log_mstar[log_mstar < -10] = -10
log_Z[log_Z < -10] = -10

### Initialize intensity map
Nx = int(args.side_length / args.angular_resolution)
Nz = len(flist) - 1
total_intensity = np.zeros((Nx, Nx, Nz), dtype=np.float32)
npix = np.array([Nx, Nx, Nz])

### Save intensity maps
output_file=f"{args.output_dir}/map_{args.side_length}sec_fmin{args.fmin}_fmax{args.fmax}_angular{args.angular_resolution}_spectral{args.R}_sigma{args.sigma}_alpha{args.z_dependence_alpha}_th{args.threshold}.h5"

with h5py.File(output_file, "w") as f:
    args_dict = vars(args)
    args_dict = {k: (v if v is not None else "None") for k, v in args_dict.items()}
    for key, value in args_dict.items():
        f.attrs[key] = value

    f.create_dataset("frequency", data=flist, compression="gzip")

    for line_name in list(line_dict.keys()):
        freq_obs = line_dict[line_name][0] / ( 1. + z_obs )
        iz = np.searchsorted(flist, freq_obs, side="right") - 1

        indices = np.array([ix, iy, iz]).T # (num_galaxies, 3)

        valid_mask = np.all((indices >= 0) & (indices < npix), axis=1)
        
        intensity_line = np.zeros((Nx, Nx, Nz), dtype=np.float32)

        if np.sum(valid_mask) > 0:
            print(f"observed frequency of {line_name} is {freq_obs[0]/1e9} - {freq_obs[-1]/1e9} [Hz]")
            print("# Found {} valid galaxies for {}".format(np.sum(valid_mask), line_name))
            print(f"observed redshift range of {line_name} is {np.min(z_obs[valid_mask])} - {np.max(z_obs[valid_mask])}")
            indices_valid = indices[valid_mask]
            log_lumi_dis_valid = log_lumi_dis[valid_mask]
            log_sfr_valid = log_sfr[valid_mask]
            log_mstar_valid = log_mstar[valid_mask]
            log_Z_valid = log_Z[valid_mask]
            z_valid = z[valid_mask]

            log_lumi_valid = calc_line_luminosity(args, z_valid, log_sfr_valid, log_mstar_valid, log_Z_valid, line_name)
            flux_valid = 10 ** ( log_lumi_valid - 2 * log_lumi_dis_valid ) / ( 4. * np.pi ) # [erg/s/cm2]

            if args.threshold < 0:
                intensity_valid = flux_valid / dflist[indices_valid[:, 2]] / Jy / ( args.angular_resolution * arcsec )**2 # [Jy/sr]
            else:
                # 1 if flux_valid > args.threshold else 0
                intensity_valid = (flux_valid > args.threshold).astype(np.float32) 
                
            np.add.at(intensity_line, (indices_valid[:, 0], indices_valid[:, 1], indices_valid[:, 2]), intensity_valid)

        total_intensity += intensity_line

        f.create_dataset("intensity_{}".format(line_name), data=intensity_line, compression="gzip")

    f.create_dataset("total_intensity", data=total_intensity, compression="gzip")

print("Intensity map saved to {}".format(output_file))

File: TNG/source/create_mock.c
#include "stdio.h"
#include "stdlib.h"
#include "math.h"
#include "float.h"
#include "hdf5.h"

#include "constant.h"
#include "para.h"
#include "proto.h"

void save_data_cube(char *fout, float *flist, float *intensity, int Nx, int Ny, int Nz);
void check_image(char *fname, float *data_cube, int Nx, int Ny, int Nz);
int binary_search(float arr[], int size, float value);
void read_filter_table(char *fname, float fmin, float fmax, float *flist, float *dflist, int *Nz);
void make_2d_gaussian_kernel(float w[], float sigma, int nw);

int main(int argc, char *argv[])
{
	int i;
	int Nx, Nz;
	if(!(argc == 11))
	{
		printf("Please input path, model id, side length [arcsec], fmin [GHz], fmax [GHz], angular resolution [arcsec], spectral resolution (R = f/df or the name of filter table if USE_FILTER_TABLE), alpha, output file id, and threshold (If threshold > 0, create galaxy number density with flux > threshold; otherwise create intensity map)\n");
		exit(1);
	}

	char path[256];
	sprintf(path, "%s", argv[1]);
	int model_id = atoi(argv[2]);
	double theta[2] = {atof(argv[3]), atof(argv[3])}; // arcsec
	double fmin = atof(argv[4]) * 1e9; // Hz
	double fmax = atof(argv[5]) * 1e9; // Hz
	double dtheta[2] = {atof(argv[6]), atof(argv[6])}; // arcsec

	double threshold = atof(argv[10]);
	if( threshold < 0 )
	{
		printf("# create intensity map\n");
	}
	else
	{
		printf("# create galaxy number density with flux > %e [erg/s/cm2]\n", threshold);
	}

	Nx = (int)( theta[0] / dtheta[0] );

	printf("# frequency: %.1f - %.1f [GHz]\n", fmin*1e-9, fmax*1e-9); 
	printf("# area: %.1f arcsec x %.1f arcsec\n", theta[0], theta[1]);
	printf("# angular resolution: %.1f arcsec x %.1f arcsec \n", dtheta[0], dtheta[1]);

	// spectral resolution //
	float *flist, *dflist;
#if defined USE_FILTER_TABLE
	char fname_filter[256];
	sprintf(fname_filter, "%s", argv[7]);
	read_filter_table(fname_filter, fmin, fmax, flist, dflist, &Nz);
	printf("# spectral resolution (GHz): %.1f \n", dflist[0]*1e-9);
#else
	double R = atof(argv[7]); // f/df	
							  
	Nz = 0;
	float fnow = fmin;
	while( fnow < fmax ) fnow += fnow / R, Nz += 1;

	flist = (float *)malloc((Nz+1) * sizeof(float));
	dflist = (float *)malloc(Nz * sizeof(float));
	for(fnow = fmin, i = 0; i < Nz; i++)
	{
		flist[i] = fnow; 
		dflist[i] = fnow / R;
		fnow += fnow / R;
	}
	flist[Nz] = fnow;
	printf("# spectral resolution (R): %.1f \n", R);
#endif

	printf("# Nx: %d, Nz: %d\n", Nx, Nz);

	// Dust-rich universe model parameters //
	double z_middle = 8; // Redshift where the sfr_factor becomes half of the maximum
	double alpha = atof(argv[8]); // Maximum factor for sfr. At highest redshift, SFR' = alpha * SFR
	printf("# z_middle: %.1f, alpha: %.1f\n", z_middle, alpha);

	// output file id //
	char fout_id[256];
	sprintf(fout_id, "%s", argv[9]);
	printf("# output file id: %s\n", fout_id);
		
	// allocate memory //
	float *intensity, *intensity_line[Nline]; 
	intensity = (float *)malloc(Nx*Nx*Nz * sizeof(float));
	for(i = 0; i < Nline; i++) intensity_line[i] = (float *)malloc(Nx*Nx*Nz * sizeof(float));

	for (i = 0; i < Nx*Nx*Nz; i++) 
	{
		intensity[i] = 0.0;
		for(int j = 0; j < Nline; j++) intensity_line[j][i] = 0.0;
	}	
	
	// read lambda and nu of each line
	double lambda[Nline], nu[Nline];
	read_line_lambda(lambda, nu);

	// read data //
	printf("### reading lightcone data %s ...\n", path);
	char fname[256];
	FILE *fp;
	sprintf(fname, "%s", path);
	fp = fopen(fname, "r");
	if(fp == NULL)
	{
		fprintf(stderr, "can not open %s\n", fname);
		exit(1);
	}

	// read header data
	double zmin, zmax, theta_lc[2];
	fscanf(fp, "%*s %*s %lf %*s %lf", &zmin, &zmax);
	fscanf(fp, "%*s %*s %*s %*s");
	fscanf(fp, "%*s %*s %lf %*s %*s %lf %*s", &theta_lc[0], &theta_lc[1]);
	for(i = 0; i < 10; i++) fscanf(fp, "%*s");
	
	printf("# redshift: %.1f - %.1f\n", zmin, zmax);
	printf("# area of lightcone: %.1f arcsec x %.1f arcsec\n", theta_lc[0], theta_lc[1]);
	double x[2], redshift, redshift_obs, log_lumi_dis, vz, rmass_half, log_sfr, log_mstar, log_Z;
	int count = 0;

	while( fscanf(fp, "%lf %lf %lf %lf %lf %lf %lf %lf %lf", &x[0], &x[1], &redshift, &log_lumi_dis, &vz, &rmass_half, &log_sfr, &log_mstar, &log_Z) != EOF )
	{
		int ix = (int)( x[0] / dtheta[0] ); 
		int iy = (int)( x[1] / dtheta[1] );
		redshift_obs = redshift + vz * 1e5 / cspeed * (1 + redshift); // redshift space; the factor 1e5 is for converting vz from [km/s] to [cm/s]
		
		// parameters for angular size of the galaxy 
		double irx = rmass_half / dtheta[0];
		int nw = (int)(irx * 3 + 1);
		float w[nw*nw];
		make_2d_gaussian_kernel(w, irx / 0.674, nw); // 50 % is within 0.674 sigma 
		
		// compute luminosity //
		double log_lumi[Nline];
		double sfr_factor = (alpha - 1) * (0.5 * ( tanh( 0.5 * (redshift - z_middle) ) + 1 )) + 1;
		log_sfr += log10(sfr_factor); // [Msun/yr]
		if( log_mstar < -10 ) log_mstar = -10;
		if( log_Z < -10 ) log_Z = -10;
		line_luminosity_CO(log_lumi, nu, pow(10.0, log_sfr), pow(10.0, log_mstar), pow(10.0, log_Z), model_id);

		// assign line fluxes to pixels //
		for(i = 0; i < Nline; i++)
		{
			double flux = pow(10.0, log_lumi[i] - 2 * log_lumi_dis) / ( 4. * M_PI ); // erg/s/cm2

			float nu_obs = nu[i] / (1.0 + redshift_obs); // [Hz]
			int iz = binary_search(flist, Nz, nu_obs);

			if( threshold < 0 ) // create intensity map
			{
				if( ix >= -nw && ix < Nx + nw && iy >= -nw && iy < Nx + nw && iz != -1)
				{
					int di[2];
					for(di[1] = -nw+1; di[1] < nw; di[1]++)
					{
						for(di[0] = -nw+1; di[0] < nw; di[0]++) 
						{
							int iw = abs(di[0]) + abs(di[1])*nw;

							int itmp = (ix+di[0])*Nx*Nz + (iy+di[1])*Nz + iz;
							float tmp = flux / dflist[iz] / Jy / (dtheta[0]*arcsec) / (dtheta[1]*arcsec) * w[iw]; // [Jy/sr]
							
							if( tmp > 0 )
							{
								intensity[itmp] += tmp; // [Jy/sr]
								intensity_line[i][itmp] += tmp; // [Jy/sr]
							}
						}
					}
				}
			}
			else // create number density map
			{
				if( flux > threshold && ix >= 0 && ix < Nx && iy >= 0 && iy < Nx && iz != -1)
				{
					//printf("%d %.4f %.4f %.4f \n", i, redshift, log_lumi[i], log10(flux));
				
					int itmp = ix*Nx*Nz + iy*Nz + iz;
					intensity[itmp] ++;
					intensity_line[i][itmp] ++;						
				}
			}
		}
		count ++;
#if defined DEBUG
	if(count > 10000) break;
#endif
	}
	printf("# number of galaxies: %d\n", count);
	fclose(fp);

	check_image("check_image.txt", intensity, Nx, Nx, Nz);

	char fout[256];
	sprintf(fout, "%s.h5", fout_id);
	save_data_cube(fout, flist, intensity, Nx, Nx, Nz);

	for(i = 0; i < Nline; i++)
	{
		sprintf(fout, "%s_%d.h5", fout_id, i);
		save_data_cube(fout, flist, intensity_line[i], Nx, Nx, Nz);
	}

	free(intensity);
	for(i = 0; i < Nline; i++) free(intensity_line[i]);
	free(flist);
	free(dflist);
	
	return 0;
}

void save_data_cube(char *fout, float *flist, float *intensity, int Nx, int Ny, int Nz)
{
	hid_t file_id = H5Fcreate(fout, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
	if( file_id < 0 )
	{
		fprintf(stderr, "can not create %s\n", fout);
		exit(1);
	}

	hid_t dataspace_id;
	hid_t dataset_id;
	
	dataspace_id = H5Screate_simple(1, (hsize_t[]){Nz+1}, NULL);
	if (dataspace_id < 0) {
        printf("Error creating dataspace\n");
        H5Fclose(file_id);
        exit(1);
    }
	dataset_id = H5Dcreate(file_id, "/frequency", H5T_NATIVE_FLOAT, dataspace_id, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
	if (dataset_id < 0) {
        printf("Error creating dataset\n");
        H5Sclose(dataspace_id);
        H5Fclose(file_id);
        exit(1);
    }

	H5Dwrite(dataset_id, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL, H5P_DEFAULT, flist);

	H5Dclose(dataset_id);
    H5Sclose(dataspace_id);

	dataspace_id = H5Screate_simple(3, (hsize_t[]){Nx, Ny, Nz}, NULL);
	if (dataspace_id < 0) {
        printf("Error creating dataspace\n");
        H5Fclose(file_id);
        exit(1);
    }
	dataset_id = H5Dcreate(file_id, "/intensity", H5T_NATIVE_FLOAT, dataspace_id, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
	if (dataset_id < 0) {
        printf("Error creating dataset\n");
        H5Sclose(dataspace_id);
        H5Fclose(file_id);
        exit(1);
    }

	H5Dwrite(dataset_id, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL, H5P_DEFAULT, intensity);

	H5Dclose(dataset_id);
    H5Sclose(dataspace_id);


    H5Fclose(file_id);

	/*
	long num[3] = {Nx, Ny, Nz};
	int status = 0;
	long fpixel[] = {1,1,1};
	int ntot = num[0] * num[1] * num[2];
	
	fitsfile *fptr;
	fits_create_file(&fptr, fout, &status);
	fits_create_img(fptr, -32, 3, num, &status);
	
	fits_write_pix(fptr, TFLOAT, fpixel, ntot, intensity, &status);
	
	fits_close_file(fptr, &status);
	*/

	printf("# output: %s\n", fout);
}

void check_image(char *fname, float *data_cube, int Nx, int Ny, int Nz)
{
	FILE *fp;
	fp = fopen(fname, "w");

	for(int i = 0; i < Nx; i++)
	{
		for(int j = 0; j < Ny; j++)
		{
			float tot = 0;
			for(int k = 0; k < Nz; k++)
			{
				tot += data_cube[i*Ny*Nz + j*Nz + k];
			}
			fprintf(fp, "%e ", tot);
		}
		fprintf(fp, "\n");
	}
	
	printf("# output: %s\n", fname);
}

// binary search
// return the index of the array where the value is located
int binary_search(float arr[], int size, float value)
{
	int left = 0;
	int right = size;

	if( value < arr[left] || value >= arr[right] )
	{
		return -1;
	}
	
	while (left <= right)
	{
		int mid = left + (right - left) / 2;
		if( value >= arr[mid] && value < arr[mid+1] )
		{
			return mid;
		}
		else if (value < arr[mid])
		{
			right = mid - 1;
		}
		else
		{
			left = mid + 1;
		}
	}
	
	return -1;

}

void read_filter_table(char *fname, float fmin, float fmax, float *flist, float *dflist, int *Nz)
{
	FILE *fp;
	float temp;

	fp = fopen(fname, "r");
	if( fp == NULL )
	{
		printf("cannot open %s\n", fname);
		exit(1);
	}

	while( fscanf(fp, "%f", &temp) != EOF ) 
	{
		if( temp > fmin && temp < fmax ) (*Nz) ++;
	}

	(*Nz) -= 1; // flist[Nz] < fmax

	fclose(fp);

	fp = fopen(fname, "r");
	
	flist = (float *)malloc((*Nz+1) * sizeof(float));
	dflist = (float *)malloc(*Nz * sizeof(float));
	int count = 0;
	while( fscanf(fp, "%f", &temp) != EOF ) 
	{
		if( temp > fmin && temp < fmax )
		{
			flist[count] = temp;
			count ++;
		}
	}

	for(int i = 0; i < *Nz; i++)
	{
		dflist[i] = flist[i+1] - flist[i];
	}

	fclose(fp);
	printf("# read filter table %s\n", fname);
}

double w_coef_2d(int i, int j)
{
	if( i*i+j*j == 0 ){ return 1.0; } // i=j=0
	else if( i*j == 0 ){ return 2.0; } // i=0 or i=j
	else{ return 4.0; } // i!=0 and j!=0
}

void make_2d_gaussian_kernel(float w[], float sigma, int nw)
{
	int i, j;
	float wtot = 0.0;

	for(i = 0; i < nw*nw; i++) w[i] = 0.0;

	for(j = 0; j < nw; j++)
	{
		for(i = 0; i < nw; i++)
		{
			double r2 = (double)( i*i + j*j );
			int itmp = i + j*nw;
			if( r2 <= (nw-1)*(nw-1) )
			{
				w[itmp] = exp( - r2 / (2.0 * sigma * sigma) );
				wtot += w_coef_2d(i,j) * w[itmp];
			}
			else
			{ 
				w[itmp] = 0.0; 
			}
		}
	}

	for(i = 0; i < nw*nw; i++) w[i] /= wtot;
}
